libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,309 +97,416 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
+    } else {
+        _attitude_target_euler_rate.x = euler_roll_rate;
+        _attitude_target_euler_rate.y = euler_pitch_rate;
+        _attitude_target_euler_rate.z = euler_yaw_rate;
     }
 
-    // Call attitude controller
-    attitude_controller_run_euler(_att_target_euler_rad, Vector3f(0.0f,0.0f,0.0f));
+    // Convert euler angle derivative of desired attitude into a body-frame angular velocity vector for feedforward
+    euler_rate_to_ang_vel(_attitude_target_euler_angle, _attitude_target_euler_rate, _attitude_target_ang_vel);
 
-    // Keep euler derivative updated
-    ang_vel_to_euler_rate(Vector3f(_ahrs.roll,_ahrs.pitch,_ahrs.yaw), _ang_vel_target_rads, _att_target_euler_rate_rads);
+    // Call quaternion attitude controller
+    attitude_controller_run_quat();
 }
 
-void AC_AttitudeControl::input_euler_rate_roll_pitch_yaw(float euler_roll_rate_cds, float euler_pitch_rate_cds, float euler_yaw_rate_cds)
+// Command an angular velocity with angular velocity feedforward and smoothing
+void AC_AttitudeControl::input_rate_bf_roll_pitch_yaw(float roll_rate_bf_cds, float pitch_rate_bf_cds, float yaw_rate_bf_cds)
 {
     // Convert from centidegrees on public interface to radians
-    float euler_roll_rate_rads = radians(euler_roll_rate_cds*0.01f);
-    float euler_pitch_rate_rads = radians(euler_pitch_rate_cds*0.01f);
-    float euler_yaw_rate_rads = radians(euler_yaw_rate_cds*0.01f);
-
-    // Compute acceleration-limited euler roll rate
-    if (get_accel_roll_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_roll_max_radss() * _dt;
-        _att_target_euler_rate_rads.x += constrain_float(euler_roll_rate_rads - _att_target_euler_rate_rads.x, -rate_change_limit_rads, rate_change_limit_rads);
-    } else {
-        _att_target_euler_rate_rads.x = euler_roll_rate_rads;
-    }
+    float roll_rate_rads = radians(roll_rate_bf_cds*0.01f);
+    float pitch_rate_rads = radians(pitch_rate_bf_cds*0.01f);
+    float yaw_rate_rads = radians(yaw_rate_bf_cds*0.01f);
 
-    // Compute acceleration-limited euler pitch rate
-    if (get_accel_pitch_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_pitch_max_radss() * _dt;
-        _att_target_euler_rate_rads.y += constrain_float(euler_pitch_rate_rads - _att_target_euler_rate_rads.y, -rate_change_limit_rads, rate_change_limit_rads);
-    } else {
-        _att_target_euler_rate_rads.y = euler_pitch_rate_rads;
-    }
+    // calculate the attitude target euler angles
+    _attitude_target_quat.to_euler(_attitude_target_euler_angle.x, _attitude_target_euler_angle.y, _attitude_target_euler_angle.z);
 
-    // Compute acceleration-limited euler yaw rate
-    if (get_accel_yaw_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_yaw_max_radss() * _dt;
-        _att_target_euler_rate_rads.z += constrain_float(euler_yaw_rate_rads - _att_target_euler_rate_rads.z, -rate_change_limit_rads, rate_change_limit_rads);
+    if (_rate_bf_ff_enabled) {
```

Comment:
Just to make sure: in the input rate methods there's no need to check *_use_input_shaping* right?